packages/flutter/lib/src/material/tabs.dart

```diff
@@ -914,6 +919,7 @@ class _TabBarState extends State<TabBar> {
     if (_controllerIsValid) {
       _controller.animation.removeListener(_handleTabControllerAnimationTick);
       _controller.removeListener(_handleTabControllerTick);
+      _controller.removeAnimationStartedListener(_handleTabControllerAnimationStarted);
```

Comment:
Ah I didn't notice there is also a scroll controller in tab bar, but it seems like we can just use the _controller.removeAnimationStartedListener(_handleTabControllerAnimationStarted); alone will be enough?
Something like..
```
_handleTabControllerAnimationStarted(int destinationIndex, Duration duration, Curve curve) {
   if (duration == null)
      duration = kTabScrollDuration;
   _scrollToIndex(destinationIndex, duration, curve);
}

void _scrollToIndex(int destinationIndex, Duration duration, Curve curve) {
    final double offset = _tabCenteredScrollOffset(destinationIndex);
    _scrollController.animateTo(offset, duration: duration, curve: curve);
  }
```

I think this should be enough to keep tabcontroller and scroll controller in sync so we don't need _handleTabControllerTick and _handleTabControllerAnimationTick.
Let me know if i miss something.

Reply:
First, we need ```_handleTabControllerAnimationTick()``` because tabbar scroll should keep in sync when user is dragging TabBarView. In that case, none of ```_handleTabControllerTick()``` or ```_handleTabControllerAnimationStarted()``` is called, so we need to do it in ```_handleTabControllerAnimationTick()```.

Second, we need ```_handleTabControllerTick()``` because TabController's ```_notifyAnimationStartedListeners()``` is not called when user drags and ballistic scroll ends. However, ```notifyListeners()``` is called, and we have to update TabBar's ```_currentIndex()``` when ballistic scroll ends, so we should do it in ```_handleTabControllerTick()```.

Reply:
I forgot about dragging yes, I agree _handleTabControllerAnimationTick is needed.

The _handleTabControllerTick though probably not needed if _notifyAnimationStartedListeners is called right? Although we need to figure out why the ballistic drag update does not drag the tabcontroller all the way to the index. Either case, _handleTabControllerAnimationStarted or _handleTabControllerAnimationTick should bring the scroll controller in tab bar to the right place?

Reply:
Yep, we wouldn't need ```_handleTabControllerTick``` if we call ```_notifyAnimationStartedListeners```. I'll change this too.